src/Npgsql/NpgsqlCommand.cs

```diff
@@ -383,6 +383,60 @@ public new NpgsqlParameter CreateParameter()
 
         #endregion
 
+        internal void DeriveParametersForQuery()
+        {
+            var connector = CheckReadyAndGetConnector();
+            using (connector.StartUserAction())
+            {
+                Log.Debug($"Deriving Parameters for query: {CommandText}", connector.Id);
+                ProcessRawQuery(true);
+
+                var sendTask = SendParseDescribe(false, false);
+
+                var isFirst = true;
+                foreach (var statement in _statements)
+                {
+                    connector.ReadExpecting<ParseCompleteMessage>();
+                    var paramTypeOIDs = connector.ReadExpecting<ParameterDescriptionMessage>().TypeOIDs;
+
+                    Debug.Assert(statement.InputParameters.Count == paramTypeOIDs.Count);
+
+                    for (var i = 0; i < paramTypeOIDs.Count; i++)
+                    {
+                        var param = statement.InputParameters[i];
+                        var derivedType = connector.TypeMapper.GetNpgsqlTypeByOid(paramTypeOIDs[i]);
+                        if(param.NpgsqlDbType != NpgsqlDbType.Unknown && param.NpgsqlDbType != derivedType)
+                        {
+                            Parameters.Clear();
+                            // Hack: I'm sure there is an established more elegant way to consume all remaining backend messages that I'm not aware of.
+                            IBackendMessage throwAwayMessage;
+                            do
+                            {
+                                throwAwayMessage = connector.ReadMessage();
+                            } while (throwAwayMessage.Code != BackendMessageCode.ReadyForQuery);
+                            throw new NpgsqlException("The backend parser inferred different types for parameters with the same name. Please try explicit casting within your SQL statement or batch or use different placeholder names.");
+                        }
+                        param.NpgsqlDbType = derivedType;
+                    }
+
+                    var msg = connector.ReadMessage();
+                    switch (msg.Code)
+                    {
+                        case BackendMessageCode.RowDescription:
+                        case BackendMessageCode.NoData:
+                            statement.Description = null;
```

Comment:
I think this line is unnecessary - when deriving parameters a command should (probably) be unprepared, so the description should already be null (it has just been created). But see comment above about what to do when deriving already-prepared commands.

Reply:
I agree. It is even somewhat dangerous not to leave statement.Description in the state it is (no matter what it is) since DeriveParameters() has no business messing around with it.
Fixed.